File: scrapers/kania_scraper.ts
/**
 * Kania Law Firm Tax Foreclosure Scraper
 * Scrapes property listings from https://kanialawfirm.com/tax-foreclosures/foreclosure-listings/
 */

import puppeteer from 'puppeteer';
import { standardizeCountyName } from '../server/utils/standardizeCounty';

interface PropertyData {
  county: string;
  address: string;
  parcelId: string;
  saleDate: string | null;
  saleTime: string | null;
  saleStatus: string;
  openingBid: number | null;
  currentBid: number | null;
  upsetBidCloseDate: string | null;
  propertyType: string;
  courtFileNumber: string;
  attorneyFileNumber: string;
  attorneyFirm: string;
  sourceUrl: string;
  sourceType: string;
}

class KaniaScraper {
  private baseUrl = 'https://kanialawfirm.com/tax-foreclosures/foreclosure-listings/';

  async scrape(): Promise<PropertyData[]> {
    const browser = await puppeteer.launch({
      headless: true,
      args: ['--no-sandbox', '--disable-setuid-sandbox', '--disable-dev-shm-usage'],
    });

    try {
      const page = await browser.newPage();
      await page.setUserAgent('Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36');
      
      console.error(`Navigating to ${this.baseUrl}...`);
      await page.goto(this.baseUrl, { waitUntil: 'networkidle2', timeout: 30000 });

      // Wait a bit for any dynamic content
      await new Promise(resolve => setTimeout(resolve, 2000));

      // Check for and handle disclaimer if present
      try {
        const checkbox = await page.$('input[type="checkbox"]');
        if (checkbox) {
          await checkbox.click();
          await new Promise(resolve => setTimeout(resolve, 500));
          
          const submitBtn = await page.$('button:has-text("Submit")');
          if (submitBtn) {
            await submitBtn.click();
            await new Promise(resolve => setTimeout(resolve, 3000));
          }
        }
      } catch (e) {
        // No disclaimer, continue
      }

      // Wait for table
      await page.waitForSelector('table', { timeout: 10000 });

      // Extract all properties from all pages
      const allProperties: any[] = [];
      let currentPage = 1;
      let hasNextPage = true;
      const MAX_PAGES = 20; // Safety limit to prevent infinite loops
      let previousPageParcelIds: Set<string> = new Set();

      while (hasNextPage && currentPage <= MAX_PAGES) {
        console.error(`Scraping page ${currentPage}...`);

        // Extract table data from current page
        const pageProperties = await page.evaluate(() => {
          const table = document.querySelector('table');
          if (!table) return [];

          const rows = Array.from(table.querySelectorAll('tr'));
          const dataRows = rows.slice(1); // Skip header

          return dataRows.map((row, rowIndex) => {
            const cols = Array.from(row.querySelectorAll('td'));
            if (cols.length < 11) return null;

            // DEBUG: Log first row to see column contents
            if (rowIndex === 0) {
              console.error(`\n=== KANIA SCRAPER DEBUG - First Row Column Contents ===`);
              cols.forEach((col, idx) => {
                console.error(`Column ${idx}: "${col?.textContent?.trim() || ''}"`);
              });
              console.error(`Total columns: ${cols.length}`);
              console.error(`=== END DEBUG ===\n`);
            }

            return {
              county: cols[0]?.textContent?.trim() || '',
              address: cols[1]?.textContent?.trim() || '',
              parcelId: cols[2]?.textContent?.trim() || '',
              saleDateTimeRaw: cols[3]?.textContent?.trim() || '',
              openingBidRaw: cols[4]?.textContent?.trim() || '',
              currentBidRaw: cols[5]?.textContent?.trim() || '',
              closeDateRaw: cols[6]?.textContent?.trim() || '',
              propertyType: cols[7]?.textContent?.trim() || '',
              courtFileNumber: cols[8]?.textContent?.trim() || '',
              attorneyFileNumber: cols[9]?.textContent?.trim() || '',
              saleStatusRaw: cols[10]?.textContent?.trim() || '',
            };
          }).filter(Boolean);
        });

        console.error(`Found ${pageProperties.length} properties on page ${currentPage}`);

        // DEBUG: Log first property to see what data was extracted
        if (currentPage === 1 && pageProperties.length > 0) {
          console.error(`\n=== KANIA SCRAPER DEBUG - First Property Data ===`);
          console.error(JSON.stringify(pageProperties[0], null, 2));
          console.error(`=== END DEBUG ===\n`);
        }

        // Stop if we got 0 properties (end of list)
        if (pageProperties.length === 0) {
          hasNextPage = false;
          console.error(`Stopping: No properties found on page ${currentPage}`);
          break;
        }

        // Detect duplicate pages (infinite loop detection)
        const currentPageParcelIds = new Set(pageProperties.map((p: any) => p.parcelId).filter(Boolean));
        
        // Check if this page has the exact same parcelIds as the previous page
        if (currentPage > 1 && currentPageParcelIds.size > 0) {
          const isDuplicate = currentPageParcelIds.size === previousPageParcelIds.size &&
                            Array.from(currentPageParcelIds).every(id => previousPageParcelIds.has(id));
          
          if (isDuplicate) {
            hasNextPage = false;
            console.error(`Stopping: Page ${currentPage} has same properties as page ${currentPage - 1} (infinite loop detected)`);
            break;
          }
        }

        // Add properties to the result (only if not duplicates)
        allProperties.push(...pageProperties);
        previousPageParcelIds = currentPageParcelIds;

        // Check if there's a next page button
        const nextButtonInfo = await page.evaluate(() => {
          // Look for the next page button (›) or the last page button (»)
          const pagination = document.querySelector('.pagination, nav[aria-label="pagination"], .page-numbers');
          if (!pagination) return { exists: false, isDisabled: false };

          // Find the next button - look for › or "Next" text
          const links = Array.from(pagination.querySelectorAll('a'));
          const nextLink = links.find(link => {
            const text = link.textContent?.trim() || '';
            return text === '›' || text.toLowerCase().includes('next') || link.classList.contains('next');
          });

          if (nextLink) {
            const isDisabled = nextLink.classList.contains('disabled') || 
                             nextLink.getAttribute('aria-disabled') === 'true' ||
                             nextLink.hasAttribute('disabled');
            return { exists: true, isDisabled, href: nextLink.getAttribute('href') };
          }
          return { exists: false, isDisabled: false };
        });

        if (nextButtonInfo.exists && !nextButtonInfo.isDisabled) {
          // Click the next button
          await page.evaluate(() => {
            const pagination = document.querySelector('.pagination, nav[aria-label="pagination"], .page-numbers');
            if (!pagination) return;

            const links = Array.from(pagination.querySelectorAll('a'));
            const nextLink = links.find(link => {
              const text = link.textContent?.trim() || '';
              return text === '›' || text.toLowerCase().includes('next') || link.classList.contains('next');
            });

            if (nextLink) {
              (nextLink as HTMLElement).click();
            }
          });

          // Wait for the page to load
          await new Promise(resolve => setTimeout(resolve, 2000));
          await page.waitForSelector('table', { timeout: 10000 });
          currentPage++;
        } else {
          hasNextPage = false;
          console.error(`No more pages. Total pages scraped: ${currentPage}`);
        }
      }

      const properties = allProperties;
      console.error(`Total properties found across all pages: ${properties.length}`);

      // Process the raw data
      const processedProperties = properties.map(prop => this.parseProperty(prop)).filter(Boolean) as PropertyData[];

      return processedProperties;

    } catch (error) {
      console.error('Error scraping Kania Law Firm:', error);
      return [];
    } finally {
      await browser.close();
    }
  }

  private parseProperty(raw: any): PropertyData | null {
    try {
      if (!raw.county || !raw.parcelId) return null;

      // Parse sale date and time
      let saleDate: string | null = null;
      let saleTime: string | null = null;
      
      if (raw.saleDateTimeRaw && raw.saleDateTimeRaw !== 'Sale date not yet set') {
        try {
          const match = raw.saleDateTimeRaw.match(/(\d{1,2})\/(\d{1,2})\/(\d{4})\s+(\d{1,2}):(\d{2}):(\d{2})\s+(AM|PM)/);
          if (match) {
            const [, month, day, year, hour, minute, second, ampm] = match;
            let hour24 = parseInt(hour);
            if (ampm === 'PM' && hour24 !== 12) hour24 += 12;
            if (ampm === 'AM' && hour24 === 12) hour24 = 0;
            
            saleDate = `${year}-${month.padStart(2, '0')}-${day.padStart(2, '0')} ${hour24.toString().padStart(2, '0')}:${minute}:${second}`;
            saleTime = `${hour}:${minute} ${ampm}`;
          }
        } catch (e) {
          // Ignore parse errors
        }
      }

      // Parse close date
      let upsetBidCloseDate: string | null = null;
      if (raw.closeDateRaw) {
        try {
          const match = raw.closeDateRaw.match(/(\d{1,2})\/(\d{1,2})\/(\d{4})/);
          if (match) {
            const [, month, day, year] = match;
            upsetBidCloseDate = `${year}-${month.padStart(2, '0')}-${day.padStart(2, '0')}`;
          }
        } catch (e) {
          // Ignore
        }
      }

      // Determine status
      let saleStatus: 'scheduled' | 'in_upset_period' | 'pending' | 'postponed' | 'sold' | 'cancelled' = 'scheduled';
      const now = new Date();
      const saleDateObj = saleDate ? new Date(saleDate) : null;
      const upsetBidCloseDateObj = upsetBidCloseDate ? new Date(upsetBidCloseDate) : null;
      
      if (upsetBidCloseDateObj && now <= upsetBidCloseDateObj) {
        // Property is in upset period (deadline exists and hasn't expired)
        // Upset period continues after sale as long as deadline hasn't passed
        saleStatus = 'in_upset_period';
      } else if (upsetBidCloseDateObj && now > upsetBidCloseDateObj) {
        // Upset period has ended
        saleStatus = 'sold';
      } else if (!saleDateObj) {
        // No sale date set yet
        saleStatus = 'scheduled';
      } else if (saleDateObj > now) {
        // Sale scheduled for future
        saleStatus = 'scheduled';
      } else {
        // Sale occurred but no upset period
        saleStatus = 'sold';
      }

      return {
        county: raw.county,
        address: raw.address,
        parcelId: raw.parcelId,
        saleDate,
        saleTime,
        saleStatus,
        openingBid: this.parseCurrency(raw.openingBidRaw),
        currentBid: this.parseCurrency(raw.currentBidRaw),
        upsetBidCloseDate,
        propertyType: raw.propertyType,
        courtFileNumber: raw.courtFileNumber,
        attorneyFileNumber: raw.attorneyFileNumber,
        attorneyFirm: 'Kania Law Firm',
        sourceUrl: this.baseUrl,
        sourceType: 'kania',
      };
    } catch (error) {
      console.error('Error parsing property:', error);
      return null;
    }
  }

  private parseCurrency(value: string): number | null {
    if (!value) return null;
    try {
      const cleaned = value.replace(/[$,]/g, '').trim();
      if (!cleaned) return null;
      return Math.round(parseFloat(cleaned) * 100); // Convert to cents
    } catch {
      return null;
    }
  }
}

export async function scrapeKania() {
  const scraper = new KaniaScraper();
  const properties = await scraper.scrape();
  
  // Transform to match PropertyData interface expected by scraperService
  return properties.map(prop => ({
    county: standardizeCountyName(prop.county || ''),
    address: prop.address,
    parcelId: prop.parcelId,
    saleDate: prop.saleDate ? new Date(prop.saleDate) : null,
    saleTime: prop.saleTime,
    saleStatus: prop.saleStatus,
    openingBid: prop.openingBid,
    currentBid: prop.currentBid,
    upsetBidCloseDate: prop.upsetBidCloseDate ? new Date(prop.upsetBidCloseDate) : null,
    propertyType: prop.propertyType,
    caseNumber: prop.courtFileNumber,
    source: 'kania',
    sourceUrl: prop.sourceUrl,
    sourceType: prop.sourceType,
    rawData: JSON.stringify(prop)
  }));
}

async function main() {
  const properties = await scrapeKania();
  console.log(JSON.stringify(properties, null, 2));
  process.exit(properties.length > 0 ? 0 : 1);
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main();
}

File: scrapers/kania_scraper.py
#!/usr/bin/env python3
"""
Kania Law Firm Tax Foreclosure Scraper
Scrapes property listings from https://kanialawfirm.com/tax-foreclosures/foreclosure-listings/
Uses Selenium for dynamic content
"""

from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.chrome.service import Service
from webdriver_manager.chrome import ChromeDriverManager
import json
import sys
from datetime import datetime
from typing import List, Dict, Optional
import time

class KaniaScraper:
    def __init__(self):
        self.base_url = "https://kanialawfirm.com/tax-foreclosures/foreclosure-listings/"
        
    def _setup_driver(self):
        """Setup Chrome driver with headless options"""
        chrome_options = Options()
        chrome_options.add_argument('--headless')
        chrome_options.add_argument('--no-sandbox')
        chrome_options.add_argument('--disable-dev-shm-usage')
        chrome_options.add_argument('--disable-gpu')
        chrome_options.add_argument('--window-size=1920,1080')
        chrome_options.add_argument('user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36')
        
        service = Service(ChromeDriverManager().install())
        driver = webdriver.Chrome(service=service, options=chrome_options)
        return driver
    
    def scrape(self) -> List[Dict]:
        """
        Scrape all property listings from Kania Law Firm website
        Returns list of property dictionaries
        """
        driver = None
        try:
            print(f"Initializing Chrome driver...", file=sys.stderr)
            driver = self._setup_driver()
            
            print(f"Navigating to {self.base_url}...", file=sys.stderr)
            driver.get(self.base_url)
            
            # Wait for page to load and check for disclaimer
            time.sleep(3)
            
            # Try to find and click "I understand" checkbox and Submit button if present
            try:
                checkbox = driver.find_element(By.CSS_SELECTOR, 'input[type="checkbox"]')
                if checkbox:
                    driver.execute_script("arguments[0].click();", checkbox)
                    time.sleep(1)
                    
                    submit_btn = driver.find_element(By.XPATH, "//button[contains(text(), 'Submit')]")
                    if submit_btn:
                        driver.execute_script("arguments[0].click();", submit_btn)
                        time.sleep(3)
            except:
                pass  # No disclaimer, continue
            
            # Wait for table to be present
            print("Waiting for property table...", file=sys.stderr)
            wait = WebDriverWait(driver, 20)
            table = wait.until(EC.presence_of_element_located((By.TAG_NAME, "table")))
            
            # Get all table rows
            rows = table.find_elements(By.TAG_NAME, "tr")
            print(f"Found {len(rows)} total rows (including header)", file=sys.stderr)
            
            properties = []
            
            # Skip header row
            for row in rows[1:]:
                try:
                    cols = row.find_elements(By.TAG_NAME, "td")
                    if len(cols) >= 10:  # Ensure we have all columns
                        property_data = self._parse_row(cols)
                        if property_data:
                            properties.append(property_data)
                except Exception as e:
                    print(f"Error parsing row: {e}", file=sys.stderr)
                    continue
            
            print(f"Successfully parsed {len(properties)} properties", file=sys.stderr)
            return properties
            
        except Exception as e:
            print(f"Error scraping Kania Law Firm: {e}", file=sys.stderr)
            import traceback
            traceback.print_exc(file=sys.stderr)
            return []
        finally:
            if driver:
                driver.quit()
    
    def _parse_row(self, cols: List) -> Optional[Dict]:
        """Parse a single table row into property dictionary"""
        try:
            # Extract text from each column
            county = cols[0].text.strip()
            address = cols[1].text.strip()
            parcel_id = cols[2].text.strip()
            sale_date_time = cols[3].text.strip()
            opening_bid = cols[4].text.strip() if len(cols) > 4 else ""
            current_bid = cols[5].text.strip() if len(cols) > 5 else ""
            close_date = cols[6].text.strip() if len(cols) > 6 else ""
            property_type = cols[7].text.strip() if len(cols) > 7 else ""
            court_file = cols[8].text.strip() if len(cols) > 8 else ""
            file_number = cols[9].text.strip() if len(cols) > 9 else ""
            
            # Skip empty rows
            if not county or not parcel_id:
                return None
            
            # Parse sale date and time
            sale_date = None
            sale_time = None
            if sale_date_time and sale_date_time != "Sale date not yet set":
                try:
                    # Format: "10/23/2025 11:00:00 AM"
                    dt = datetime.strptime(sale_date_time, "%m/%d/%Y %I:%M:%S %p")
                    sale_date = dt.strftime("%Y-%m-%d %H:%M:%S")
                    sale_time = dt.strftime("%I:%M %p")
                except:
                    # Try without time
                    try:
                        dt = datetime.strptime(sale_date_time, "%m/%d/%Y")
                        sale_date = dt.strftime("%Y-%m-%d")
                    except:
                        pass
            
            # Parse close date (upset bid deadline)
            upset_close_date = None
            if close_date and close_date != "":
                try:
                    dt = datetime.strptime(close_date, "%m/%d/%Y")
                    upset_close_date = dt.strftime("%Y-%m-%d")
                except:
                    pass
            
            # Determine sale status
            sale_status = "scheduled"
            if upset_close_date:
                sale_status = "in_upset_period"
            elif not sale_date:
                sale_status = "scheduled"
            
            # Parse bid amounts (convert to cents)
            opening_bid_cents = self._parse_currency(opening_bid)
            current_bid_cents = self._parse_currency(current_bid)
            
            return {
                "county": county,
                "address": address,
                "parcelId": parcel_id,
                "saleDate": sale_date,
                "saleTime": sale_time,
                "saleStatus": sale_status,
                "openingBid": opening_bid_cents,
                "currentBid": current_bid_cents,
                "upsetBidCloseDate": upset_close_date,
                "propertyType": property_type,
                "courtFileNumber": court_file,
                "attorneyFileNumber": file_number,
                "attorneyFirm": "Kania Law Firm",
                "sourceUrl": self.base_url,
                "sourceType": "kania",
            }
        except Exception as e:
            print(f"Error parsing property row: {e}", file=sys.stderr)
            return None
    
    def _parse_currency(self, value: str) -> Optional[int]:
        """Convert currency string like '$3,400.00' to cents"""
        if not value or value == "":
            return None
        try:
            # Remove $ and commas
            cleaned = value.replace('$', '').replace(',', '').strip()
            if not cleaned:
                return None
            # Convert to float then to cents
            return int(float(cleaned) * 100)
        except:
            return None


def main():
    """Main entry point for scraper"""
    scraper = KaniaScraper()
    properties = scraper.scrape()
    
    # Output JSON to stdout
    print(json.dumps(properties, indent=2))
    
    # Return exit code based on success
    return 0 if properties else 1


if __name__ == "__main__":
    sys.exit(main())
